dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+            }
+            JsonBinary::appendSIMDJsonElem(write_buffer, json_elem.value_unsafe());
+
+            writeChar(0, write_buffer);
+            offsets_to[i] = write_buffer.count();
+            data_from->next();
+        }
+    }
+
+    static void doExecuteForOthers(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const std::unique_ptr<IStringSource> & data_from,
+        size_t size)
+    {
+        size_t reserve_size = (size * (1 + 1)) + data_from->getSizeForReserve();
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to, reserve_size);
+        for (size_t i = 0; i < size; ++i)
+        {
+            const auto & slice = data_from->getWhole();
+            JsonBinary::appendStringRef(write_buffer, StringRef{slice.data, slice.size});
+            writeChar(0, write_buffer);
+            offsets_to[i] = write_buffer.count();
+            data_from->next();
+        }
+    }
+
+private:
+    const tipb::FieldType * input_tidb_tp = nullptr;
+    const tipb::FieldType * output_tidb_tp = nullptr;
+    TiDB::TiDBCollatorPtr collator = nullptr;
+};
+
+class FunctionCastTimeAsJson : public IFunction
+{
+public:
+    static constexpr auto name = "cast_time_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastTimeAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    void setInputTiDBFieldType(const tipb::FieldType & tidb_tp_) { input_tidb_tp = &tidb_tp_; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if unlikely (!arguments[0]->isMyDateOrMyDateTime())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        const auto & from = block.getByPosition(arguments[0]);
+        if (checkDataType<DataTypeMyDateTime>(from.type.get()))
+        {
+            doExecute<DataTypeMyDateTime, false>(data_to, offsets_to, from.column);
+        }
+        else if (checkDataType<DataTypeMyDate>(from.type.get()))
+        {
+            // In raw function test, input_tidb_tp is nullptr.
+            bool is_timestamp = (unlikely(input_tidb_tp == nullptr)) || input_tidb_tp->tp() == TiDB::TypeTimestamp;
```

Comment:
Why a `MyDate` type in TiFlash might be `Timestamp` type in TiDB?

Reply:
> Why a MyDate type in TiFlash might be Timestamp type in TiDB?
my mistake, should be MyDateTime.

Reply:
> And why is_timestamp is by default true if input_tidb_tp is nullptr?
There is no specific reason, it's just that in the raw_function_test, the tipb::FieldType will be empty, just setting is_timestamp = true.

Reply:
So it should never be nullptr in real env?